packages/app-defaults/package.json

```diff
@@ -41,8 +41,8 @@
     "@material-ui/icons": "^4.9.1"
   },
   "peerDependencies": {
-    "react": "^16.13.1 || ^17.0.0",
-    "react-dom": "^16.13.1 || ^17.0.0",
+    "react": "^17.0.0 || ^18.2.0",
+    "react-dom": "^17.0.0 || ^18.2.0",
```

Comment:
We'll need to consider the impact of dropping support for react 16. Honestly I think we need to just go for it, because MUI v5 doesn't support 16 either and we want to go there soon.

I do however feel it might be best to have a release with very strong guidance towards 17, perhaps even a warning for apps that are still on 16. It all depends on the impact really. Do you have a sense for what might break if forced to move to from 16 to 17?

Reply:
I haven't tried breaking between 16 & 17, but the amount of breaking changes [is minimal](https://legacy.reactjs.org/blog/2020/10/20/react-v17.html#other-breaking-changes). 
Most of the APIs are the same from what I can tell. It's the React 16/17 --> 18 migration which is much more breaking. 

I can also split this PR up to first migrate off of React 16 to 17. This may violate the sliding window upgrade policy that Backstage has.

Reply:
Ended up with #17752, which takes care of the v16 deprecation path. With that in place we can keep the focus of this PR to be the 16+17 -> 17+18 shift, which we'll be able to ship after the release next week.